### ISSUE TITLE
MySQL 8 compatibility issue with Product data stores

### ISSUE THREAD

**Issue by jkohlbach** [woocommerce/woocommerce#28417]
In includes/data-stores/class-wc-product-data-store-cpt.php there's a line 1617:

$variation_query = $wpdb->prepare( ' OR ( wc_product_meta_lookup.sku = "" AND parent_wc_product_meta_lookup.sku LIKE %s ) ', $like );

Although some folks think its OK to use "" and '' interchangeably in SQL syntax, correct syntax is using SINGLE quotes.

> ANSI_QUOTES
> Treat " as an identifier quote character (like the ` quote character) and not as a string quote character. You can still use ` to quote identifiers with this mode enabled. With ANSI_QUOTES enabled, you cannot use double quotation marks to quote literal strings because they are interpreted as identifiers.

This causes issue with MySQL 8 in certain configurations.

**Prerequisites (mark completed items with an [x]):**
- [x] I have have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
You should be consistent with other code and use single quotes on all SQL statements. Double quotes is incorrect syntax.

One symptom of this bug (it's possible that it affects other things) is you can't search for products when editing an Order on the backend:

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
Steps to reproduce the bug:
Prerequisite: be on MySQL 8 with ANSI_QUOTES enabled
1. Make an manual order on the backend
2. Attempt to add a product, you'll get an error like:

`[26-Nov-2020 06:38:09 UTC] WordPress database error Unknown column '' in 'where clause' for query SELECT DISTINCT posts.ID as product_id, posts.post_parent as parent_id FROM wp_posts posts
                         LEFT JOIN wp_wc_product_meta_lookup wc_product_meta_lookup ON posts.ID = wc_product_meta_lookup.product_id
                          LEFT JOIN wp_wc_product_meta_lookup parent_wc_product_meta_lookup
                         ON posts.post_type = 'product_variation' AND parent_wc_product_meta_lookup.product_id = posts.post_parent
                        WHERE posts.post_type IN ('product','product_variation')
                         AND (  ( ( posts.post_title LIKE '%SEARCHTEXT%') OR ( posts.post_excerpt LIKE '%SEARCHTEXT%') OR ( posts.post_content LIKE '%WWPP%' ) OR ( wc_product_meta_lookup.sku LIKE '%SEARCHTEXT%' )  OR ( wc_product_meta_lookup.sku = "" AND parent_wc_product_meta_lookup.sku LIKE '%SEARCHTEXT%' ) ))
                         AND posts.post_status IN ('private','publish')

                        ORDER BY posts.post_parent ASC, posts.post_title ASC
                         LIMIT 30
                         made by do_action('wp_ajax_woocommerce_json_search_products_and_variations'), WP_Hook->do_action, WP_Hook->apply_filters, WC_AJAX::json_search_products_and_variations, WC_AJAX::json_search_products, WC_Data_Store->__call, WC_Product_Data_Store_CPT->search_products
`

**Expected behavior**
MySQL 8 should be able to be configured this way.

**Actual behavior**
Error is thrown as above

**WordPress Environment**
Confirmed on MySQL 8 with ANSI_QUOTES enabled.

**Comment by rrennick** [comment]
Hi @jkohlbach ,

Thank you for taking the time to report this bug, we really appreciate your help. I can reproduce it on my end. The steps I used were

- PHP 7.4.8 (8.0 RC will also work)
- edit `my.cnf` to add
```
sql_mode=ANSI_QUOTES
```
- Create a WP DB user for MySQL 8.0
```
mysql> CREATE USER 'wcuser'@'%' IDENTIFIED WITH mysql_native_password BY 'REDACTED';
Query OK, 0 rows affected (0.02 sec)

mysql> FLUSH PRIVILEGES;
Query OK, 0 rows affected (0.01 sec)

mysql> SELECT user,authentication_string,plugin,host FROM mysql.user WHERE user = 'wcuser';
+--------+-------------------------------------------+-----------------------+------+
| user   | authentication_string                     | plugin                | host |
+--------+-------------------------------------------+-----------------------+------+
| wcuser | *REDACTED | mysql_native_password | %    |
+--------+-------------------------------------------+-----------------------+------+
1 row in set (0.00 sec)
```
- Update `wp-config.php` with `wcuser` credentials
- Update to MySQL 8.0
- Restart the MySQL service
- Confirm MySQL 8.0 on SSR
- Attempt to add product to order

We’re going to solve this bug as soon as possible. Stay tuned for updates.

**Comment by rodrigoprimo** [comment]
Thanks for the report, @jkohlbach. I opened #28468 to address this issue.

You mention specifically that this is a problem with MySQL 8, but after a quick research, it seems to me that the ANSI_QUOTES mode has been around for a long time (for example, see this WP core ticket from 2005: https://core.trac.wordpress.org/ticket/1498) and I couldn't find any related changes in MySQL 8. Am I missing something?

**Comment by jkohlbach** [comment]
@rodrigoprimo yeah looks like the mode has been around for a bit. I think it's becoming more common to have this turned on in MySQL 8 which might be why I encountered it (I first encountered it on a Digital Ocean box and perhaps their default config is more strict)

Appreciate you pushing the fix through all the same! I just wanted to make sure it's solid for mysql8 and folks are aware they shouldn't be using double quote.
